### ISSUE TITLE
Monthly Trends

### ISSUE THREAD

**Issue by ultimike** [communityinclusion/escoach#58]
This is an offshoot of [task 56](https://github.com/communityinclusion/escoach/issues/56#issuecomment-911827703).

We need to validate the query modifications for monthly (primary) and quarterly trends.

See [this comment on  task 20](https://github.com/communityinclusion/escoach/issues/20#issuecomment-851572780) for sample queries.

**Comment by ultimike** [comment]
@escoach Let's do the same thing with monthly trends that we did for the rest of the other queries and data tables over the past few weeks. 

I've gotten the ball started with the following example:

- Who=no selection
- Where=no selection
- What=(Getting to know a job seeker)Observing a job seeker=answer537=11838, Participating=answer537=11899, Informational interviews=answer537=12115
- Time frame = Monthly trends

I manually ran the query based on https://github.com/communityinclusion/escoach/issues/20#issuecomment-851572780 and output the data on the "Monthly trends" tab of [my Google Sheet.](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=1862476710)

Currently, my manual output doesn't totally match the /dashboard output for this scenario but before we even start looking at code, I'd love for you to create the reference data table so that we know what we're aiming for. 

Does that work for you?

thanks,  
-mike

**Comment by escoach** [comment]
@ultimike A few questions & comments before I dig deeper into this. What timeframe should we use and which one did you pick for your Google sheet? 

For example:
2. "All" data runs from May 2020 to May 2021 (Excluding "Observer"). I counted 318 total responses. You have 353. Did you exclude the "Observer"? If yes, am I missing something?
3. "My team" data runs from April 2021 to July 2021. I counted 68 like you. Nice!
4. "Me" data runs from  July 2020 to July 2021 (13 months). I have 122 if I count July 2020 to June 2021. Or I have 133 if I count August 2020 to July 2021. You have 121. Strange?

I think that we should pick the period August 2020 to July 2021 (most recently available 12 months) for consistency?
This leads me to the second question.  How are we going to address the fact that ideally, the table shows the last 12 months only? For example, right now the table should show September 2020 to August 2021. Next October it will show October 2020 to September 2021. Is this even possible without major complications? What options do we have?

**Comment by ultimike** [comment]
@escoach:

- The data in the Google Sheet has nothing to do with the Observer task.  
- The monthly trends query is only looking at data from the past 1 year - this is what I believe we discussed doing previously.  
- The query will always only show data from last 12 months - in layman's terms, the query is doing something like this: "show me data from the last 1 year, grouped by month". 
- Limiting the "what" data to only the level 1 items would mean that we'd need to modify the user interface so that users cannot select level 2 items if the "monthly" time frame is selected, correct?

-mike

**Comment by escoach** [comment]
@ultimike Thank you for your clarifications. Let's hold off a few more days on the trend data. I need clarifications from the team here before we move forward.

**Comment by ultimike** [comment]
Relocating the discussion from #56 to here:

> @ultimike Below is @escoach 's specification of what is needed for the monthly trends display. Let us know if this adds significantly to hours or if there are difficulties with this.
> 
> Here is what to ask Mike in regard to the trend data tables:
> Generate a monthly trend data as visualized in Table 1 in this Google sheet: https://docs.google.com/spreadsheets/d/1xKtCi2jgxygtJyhGS4qTN7I6rqtiGHx-LpsuW7877lk/edit#gid=1970680850 (Tab "Trends0916", See cell A2:F17) Note:
> 
> I included the most recent 12 months' worth of data: from August 1, 2020 to July 31, 2021.
> 
> The time value for each month is an average of 3 months. For ex., the time values in October is the average of Aug, Sep, and Oct.
> 
> If it is not too complicated, include the n selected activities per each month for the "me" data (Column F in the sheet.)
> 
> All formulas are available in the cells of the sheet.
> 
> Remember to exclude the "Observers"
> 
> Once we agree on this format, I will be happy to generate a few more tables for other types (e.g. Who).
> 
> PS: The quarterly tables will be very easy to generate. for example, the data for the quarter Oct 2020-Dec 2020 is A10:F10.

@escoach @thinkcollege This is definitely going to complicate things a bit compared to our original approach to the trends data. 

Before I can make much headway into figuring out the queries, I need to know a couple of things:

1. In the sample Google Sheet data, what user and provider were used to generate the data? If I am to replicate it, this is vital information. I assume it is Alberto's user, but is his provider set to ICI or Observer? (I think ICI, please confirm.) 
2. Can I get an exact copy of the `surveycampaign_results` table that was used to generate the Google Sheet data? We need to make sure we're working from the same data.

thanks,  
-mike

**Comment by thinkcollege** [comment]
@escoach  In response to Mike's questions above, do you know how you got the sample data in the Google Sheet?  Was it from the live site?  Do you know when it was generated?  That would help me find the data table it came from.

**Comment by escoach** [comment]
Hi @thinkcollege and @ultimike, thank you for the question. The data and the user and provider are the same that we used for the "up-to-date" tables. To be precise and minimize confusion:

**Question1:** 
"Me" data = alberto.migliore@umb.edu
"My team" = The Institute for community inclusion. However, like in the "up-to-date" tables, "The Institute for Community Inclusion" is an "Observer" and, therefore, should be removed from the "All" data.

**Question2:** 
I used the same data as before. But as you requested, and to be exclude misunderstanding, I added the data in the tab "Data" in the [Google sheet](https://docs.google.com/spreadsheets/d/1xKtCi2jgxygtJyhGS4qTN7I6rqtiGHx-LpsuW7877lk/edit#gid=1970680850)

Let me know if you have any questions. Between 1 and 5 I will be in a long meeting with limited ability to respond. But will do if I can.

**Comment by ultimike** [comment]
@escoach @thinkcollege I think we need to get all sync'd up on the data. 

1. My local dataset that I use for development has a latest `surveycampaign_result` record dated 2021-06-03.  
2. [Alberto's Google Sheet ("Data" tab)](https://docs.google.com/spreadsheets/d/1xKtCi2jgxygtJyhGS4qTN7I6rqtiGHx-LpsuW7877lk/edit#gid=1798624280) has a latest record from 2021-07-21.  
3. I'm not sure what the latest record on the Dev site is (I suspect it is the same as my local).  

@thinkcollege Can we update the Dev site with up-to-date data? Then, send me a copy of the table to import to my local? Also, send a copy to @escoach to update his Google Sheet with? This is the most sure-fire way of making sure we're not chasing our tails...

If we do update Dev with up-to-date data, do we need to run the query to transform all ICI data to "Observer"?

`UPDATE surveycampaign_results SET provider = 'Observer' WHERE provider = 'Institute for Community Inclusion';`

-mike

**Comment by thinkcollege** [comment]
I updated the sp-escoach.communityinclusion.org site with the latest surveycampaign_results data from the live site, es-coach.org. Attached is a mysqldump file of the live surveycampaign_results table.
[surveycampaign_results.zip](https://github.com/communityinclusion/escoach/files/7204769/surveycampaign_results.zip)
I sent Alberto a CSV version of that table.  The change to Observer has already been run on the development site.

**Comment by ultimike** [comment]
@thinkcollege Thank you.

I have imported the data into my local site. The most recent `date_submitted` value is 2021-09-03.

It looks like there is 1 record where the provider is `Institute for Community Inclusion` and 132 records where the provider is `Observer`.  

@escoach Please do not modify the data on your end when updating your Google Sheet, that way we are sure to be comparing apples to apples.

I have updated my [Google Doc (Monthly Trends tab)](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=1862476710) with query results from this update data. **But, I have not yet implemented the "average of 3 months" bit yet.**  How do we handle the first two months with this? For example, for January 2021, do we average December 2021, November 2021, and January 2021?

Or, do we average November 2020, December 2020, and January 2021? If so, we need to query for 14 months, but only use the oldest 2 months for averaging? This gets very tricky very quickly...

thanks,  
-mike

**Comment by thinkcollege** [comment]
I just looked at the sql file I sent and the most recent date is 9/20/21, same as the surveycampaign_results table on the development site.  Here it is agai
[surveycampaign_results_2.zip](https://github.com/communityinclusion/escoach/files/7206148/surveycampaign_results_2.zip)
n.

**Comment by escoach** [comment]
Thank you @ultimike and thanks @thinkcollege for the follow-up about the most recent date. Below are my responses to your questions and a few other points for consideration to make sure we are looking at the same data:

- The file I am looking at has 1,151 records in total
- The valid records are 768 (i.e. What 483 > 0). We should use only these valid records for the table.
- 383 records are not usable (Mostly 482 = 11760 out of the office or the survey is incomplete) 
- For January 2021, we would average January 2021, December 2020, and November 2020. But if this is complicated, let's talk. 

Would it be ok to include in the table the most recent 12 complete months, to mimic reality? I mean, September 1 2020 to August 31 2021? Not a big deal, I can create the January-December table if that is easier. 

Alberto

**Comment by escoach** [comment]
@ultimike check out the new Tab "[Trends092121](https://docs.google.com/spreadsheets/d/1xKtCi2jgxygtJyhGS4qTN7I6rqtiGHx-LpsuW7877lk/edit#gid=1136224461)" in the Google sheet.

I could not replicate exactly the January to December period shown in your table. The data available runs from May 14 2020 to September 20, 2021. So If I start with January (I presume it's January 2021) I have data only through September 2021, not through December.

Instead, I picked the months corresponding to what users would see if they would access the dashboard today: September 2020 to August 2021. 

September and October 2020 are special cases because we do not have 3 months' worth of data. But the issue occurs only at the onset of the project because the system does not have complete data yet (Or when a new user registers). My inclination would be to run the average with whatever amount of data are available at that time. But if this overcomplicates things, let @thinkcollege know. Our team will explore if there are other simpler options that will still add value to users' experience.

**Comment by ultimike** [comment]
@thinkcollege I see what happened - I was looking at the last row of the file you sent me - the date is `2021-09-03`, but the insert queries are not sorted by date. I see the `2021-09-20` record that you are referring to. I think we're working with the same data. 😄 

Following up on @escoach's comments:

1. Agreed, there are 1,151 rows of data.
2. Agreed, there are 768 rows where answer483 is not null.  
3. I don't understand how you determined that 383 records are not usable using `Mostly 482 = 11760 out of the office or the survey is incomplete`. Is this something new? I don't believe any of our other queries are disqualifying records in this manner. 1,151-768=383, so I'm going to assume that the logic we're already using (answer483 is not null) is sufficient. 
4. Yes, I'm pretty sure that using data from earlier than the previous 12 months is going to complicate things. It means that we'll have to expand the query to include 14 months of data without introducing any adverse affects (which I'm not sure how to do at this point). For now, I'd rather just focus on making sure we are on the same page with unaveraged data before we go down the road of 3-month averages. It would be really helpful if you updated your Google Sheet to remove the 3-month average for now, know what I mean? In the interim, I've copied your "Trends092121" sheet into [my Google Sheet's "Monthly Trends - Alberto data" tab](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=1321166921) where I have removed the 3 month average.
5. I have modified the query (on my local - not ready for testing on sp-escoach yet) to only include the most recent 12 complete months. 
6. I think that maybe there is a bug in your spreadsheet, @escoach. Looking at data from January, please take a look at the formula in cell J10: to calculate the value you're subtracting the `Team` value from the `All (Includes "Observers")` value. I don't think the `Team` value is relevant in this calculation. Same for cell O10. This calculation is used for all other months as well. Is this correct?

-mike

**Comment by ultimike** [comment]
@escoach Something strange is going on.

I tweaked the calculation I use to generate my table on [my Monthly Trends Google Sheet](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=1862476710).

My "All (hrs/day)" values match your Google Sheet's values from February - August and again from October - December. **January and September are outliers.** Could there be some bug in the values in columns K-N and Q-T on your Google Sheet for January and September?

Furthermore, all of the "Me (hrs/day)" data matches between our two sheets.

But, none of the "Team (hrs/day)" data matches. This might have something to do with how one of is us counting team data. I am assuming that the "me" has the ICI provider, which is different than the "Observer" provider. I suspect that in your Google Sheet, you are assuming that "me" has the "Observer" provider.

I think we are very close...

-mike

**Comment by escoach** [comment]
@ultimike strange. Here is what I used for the Google sheet:
Me = alberto.migliroe@umb.edu 
My team = "Observer". 

You are right that alberto.migliore@umb.edu is associated with multiple providers. Specifically, of the 768 total valid records, I see 163 records by alberto.migliore@umb.edu (Me data). Of these, the provider is Testing (n=21), Observer (N= 53), and No Provider (N=89). I do not see ICI provider associated with alberto.migliore@umb.edu. 

I used "Observer" because it's the equivalent of "The Institute for community inclusion" that we used in the previous 'up-to-date' tables. And it is how I will be categorized in the data with every new daily survey. 
 
Understanding the roots of this discrepancy might address also the other issue with the January and September data?
I am glad that the "All" values match, getting there.

**Comment by ultimike** [comment]
@escoach Ok - that helped. I was using the wrong provider in my query. Changing that and all of the "My team" data now matches your Google Sheet with the exception of **September**. 

January and September of the "All" data does still not match. 

All of the "Me" data still matches.

I am completely at a loss as to why we are having issues with January and September.

Here is the query I'm using:

```
SELECT
    count(*) AS TotalAll,
       sum(case when ((answer483 in (11658, 11659, 11661) )) then 1 else 0 end) AS SelectedAll,
       sum(case when ((answer483 in (11658, 11659, 11661) )) AND email = 'alberto.migliore@umb.edu' then 1 else 0 end) AS SelectedMe,
       sum(case when ((answer483 in (11658, 11659, 11661) )) AND provider = 'Observer' then 1 else 0 end) AS SelectedProvider,
       sum(case when ((answer483 in (11658, 11659, 11661) )) AND provider = 'Observer' then 1 else 0 end) AS SelectedObserver,
    sum(case when (email = 'alberto.migliore@umb.edu') then 1 else 0 end) AS TotalMe,
    sum(case when (provider = 'Observer') then 1 else 0 end) AS TotalProvider,
    sum(case when (provider = 'Observer') then 1 else 0 end) AS TotalObserver,
       sum(case when ((answer483 in (11658, 11659, 11661) )) then 1 else 0 end) AS nAll,
       sum(case when ((answer483 in (11658, 11659, 11661) )) AND provider = 'Observer' then 1 else 0 end) AS nProvider,
       sum(case when ((answer483 in (11658, 11659, 11661) )) AND provider = 'Observer' then 1 else 0 end) AS nObserver,
       sum(case when ((answer483 in (11658, 11659, 11661) )) AND email = 'alberto.migliore@umb.edu' then 1 else 0 end) AS nMe,
    MONTH(date_submitted) AS month
FROM surveycampaign_results
WHERE (answer483 IS NOT NULL)
/* Previous 12 completed months */
AND (date_submitted > DATE_SUB(DATE_SUB(LAST_DAY(NOW()),INTERVAL DAY(LAST_DAY(NOW()))- 1 DAY), INTERVAL 1 YEAR))
GROUP BY MONTH(date_submitted);
```

If you could shed any light on things, I would be most appreciative.

-mike

**Comment by escoach** [comment]
@ultimike I think I found out what's going on.
For January, it was my mistake. I counted 6 Observers among the "Not selected" when in fact there were none. Now my google sheet data for January "All" (B9) matches your table (B4, i.e. 2:00 for All).

As for the September data, it looks like your query is adding the September 2020 data and the September 2021 data. The table should include only September 2020, which includes only 14 total records, not 82 as shown in cell A29 on your spreadsheet.  

I think that's it.

**Comment by ultimike** [comment]
@escoach Yes - I think you are correct. I found and fixed the issue in my query so that it is only including records between 9/1/2020 and 8/30/2021 - I've updated my Google Sheet and I think all numbers are matching. Feel free to confirm.

I believe the next step is for me to chat with @mgcobblestone about implementation, including discussing the feasibility of 3 month averaging.

-mike

**Comment by ultimike** [comment]
@mgcobblestone Here's the legend we discussed:

```
 n selected activities: All (No observers) = SelectedAll - SelectedObserver
 n selected activities: All (Includes observers) = SelectedAll
 n selected activities: Team = SelectedProvider
 n selected activities: My data = SelectedMe
 n NOT selected: All (No observers) = (TotalAll - SelectedAll) - (TotalObserver - SelectedObserver)
 n NOT selected: All (Includes observers) = TotalAll - SelectedAll
 n NOT selected: Team = TotalProvider - SelectedProvider
 n NOT selected: My data = TotalMe - SelectedMe
 ```

-mike

**Comment by mgcobblestone** [comment]
@ultimike , I've updated the calculation based on your spreadsheet.  I had to manually calculate the date range because I was getting no results at all using the date_sub stuff.

One question:  Will Selected Activities - Up-to-Date use the same calculations?

**Comment by ultimike** [comment]
What do you mean "no results at all"? The query wasn't returning anything?

Here's an updated `surveycampaign_results` table that you can import to you local (is it only the one table) so we're all working with the same data: [surveycampaign_results-live-sept-21.sql.zip](https://github.com/communityinclusion/escoach/files/7239237/surveycampaign_results-live-sept-21.sql.zip)

No - we're not changing any of the other tables/results/calculations - this is just for the Monthly Trends stuff.

Testing on my local - the "Me" and "My Team" data looks good, but the "All" data doesn't match up with [my Google Sheet (Monthly Trends tab)](https://docs.google.com/spreadsheets/d/13T2JrHmX-xpxbkB5wk5z9Br4V5hauoWlL4gy0pg7ZXA/edit#gid=1862476710). (Update your `surveycampaign_results` table so you can compare your local with my Google Sheet as well.)

Also - getting some minor PHP Notices on /dashboard default display (before changing any options):

```
Notice: Undefined index: SelectedAll in Drupal\survey_dashboard\Service\QueryBuilder->calculateHrs() (line 418 of modules/custom/survey_dashboard/src/Service/QueryBuilder.php).
Notice: Undefined index: SelectedObserver in Drupal\survey_dashboard\Service\QueryBuilder->calculateHrs() (line 419 of modules/custom/survey_dashboard/src/Service/QueryBuilder.php).
```

-mike

**Comment by ultimike** [comment]
As discussed previously, the 3-month average stuff would complicate things on our end quite a bit - to what level we're not sure yet, but we know that it will force us to run an additional query. I would prefer if we get the Monthly Trends stuff (and possibly Quarterly Trends) working without it first, then see where we are before you and I decide if we want to tackle that as well.

-mike

**Comment by escoach** [comment]
@ultimike 
I reviewed the http://sp-escoach.communityinclusion.org/dashboard and I confirm that the data matches our google sheets.

**Question**: "_Determine which values go in the "Total respondents" row for this case (and update Google Sheets)._"
**Answer**: Instead of a "Total responses" row, could you do a "Total responses" column on the right side of the table? It should count the total number of responses available for the "Me" data, each month. See column [H in the Google sheet, Tab "Trends092121"](https://docs.google.com/spreadsheets/d/1xKtCi2jgxygtJyhGS4qTN7I6rqtiGHx-LpsuW7877lk/edit#gid=1136224461).  


**Question**: _Generate a new test data table with different inputs to further validate the code (perhaps with a single who or what option selected)._
**Answer**: I will share something by the end of the day today.

**Question**: _Are "Who=Any interaction" or "Where=Any place" valid selections for Monthly Trends? If so, we'll need new test data tables in Google Sheets to validate with._
**Answer**: "Who = Any interaction" is the same as "Who = no selection". The same applies to the "Where = Anyplace" is the same as "Where = No selection". Rationale: With the "up-to-date" tables we needed the "Any interactions" and "Anyplace" queries to generate the summary tables for Who and Where. But for the trend data, the summary tables would be complicated and probably not very useful. 

Finally, you probably plan to address it later, but the table should start with the first month of the 12 month period (i.e. September 2020, not January 2021).

**Comment by ultimike** [comment]
"Total respondents" stuff moved to #68 

"Any interactions" or "Any place" stuff moved to #69

Table row starting point moved to #70

I vote we mark this issue as "complete" in lieu of smaller, more focused, and more easily followable issues.

@escoach If you agree, please mark this as closed.

thanks,
-mike
